File: algo/39.py
from typing import List

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []

        def dfs(csum: int, index: int, path: List[int]):
            if csum < 0:
                return
            if csum == 0:
                result.append(path)
                return
            for i in range(index, len(candidates)):
                dfs(csum - candidates[i], i, path + [candidates[i]])
        
        dfs(target, 0, [])
        return result

a = Solution()

candidates = [2,3,6,7]
target = 7
print(a.combinationSum(candidates, target))

candidates = [2,3,5]
target = 8
print(a.combinationSum(candidates, target))

candidates = [2]
target = 1
print(a.combinationSum(candidates, target))

File: old/algo/39.kt
class Solution {
    fun combinationSum(candidates: IntArray, target: Int): List<List<Int>> {
        val result = mutableListOf<List<Int>>()
        fun visit(start: Int, sum: Int, path: List<Int>) {
            if (sum == target) {
                result.add(path)
                return
            } else if (sum > target) {
                return
            }
            for (i in start..candidates.lastIndex) {
                val candidate = candidates[i]
                visit(i, sum + candidate, path + candidate)
            }
        }
        visit(0, 0, listOf())
        return result
    }
}

//val candidates = intArrayOf(2,3,6,7)
//val target = 7

//val candidates = intArrayOf(2,3,5)
//val target = 8

val candidates = intArrayOf(2)
val target = 1
println(Solution().combinationSum(candidates, target))
